### PR TITLE
feat(filterable-select): add list action button (FE-3034)

### PR DESCRIPTION
**Pull request by grabkowski**
### Proposed behaviour
A new customizable button in Select List that triggers a user defined action.

Screenshot 2020-10-09 at 10 51 28

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions
1. npm start
2. open http://localhost:9001/?path=/docs/design-system-select-filterable--with-action-button

### CODE REVIEW

File: src/components/select/filterable-select/filterable-select.spec.js [hunk 1]
```diff
@@ -369,23 +370,88 @@ describe('FilterableSelect', () => {
...
+  describe('when the onKeyDown prop is passed', () => {
+    const expectedEventObject = {
+      key: 'ArrowDown'
+    };
 
-describe('when the onKeyDown prop is passed', () => {
-  const expectedEventObject = {
-    key: 'ArrowDown'
-  };
+    it('then when a key is pressed, that prop should be called with expected values', () => {
```
Comment (dawidzarzycki) [inline src/components/select/filterable-select/filterable-select.spec.js:379]: I think that we should to check also if the list appears after using `arrowDown`

File: src/components/select/filterable-select/filterable-select.spec.js [hunk 2]
```diff
@@ -369,23 +370,88 @@ describe('FilterableSelect', () => {
...
+
+      wrapper.find('input').simulate('keyDown', expectedEventObject);
+
+      expect(onKeyDownFn).toHaveBeenCalledWith(expect.objectContaining({
+        ...expectedEventObject
+      }));
+    });
+  });
+
+  describe('when the lisActionButton prop is provided', () => {
```
Comment (dawidzarzycki) [inline src/components/select/filterable-select/filterable-select.spec.js:391]:
```suggestion
  describe('when the listActionButton prop is provided', () => {
```

File: src/components/select/filterable-select/filterable-select.spec.js (hunk 1 above)
Comment (grabkowski) [reply src/components/select/filterable-select/filterable-select.spec.js:379]: that check is already there on line 80:
https://github.com/Sage/carbon/pull/3271/files/316cd4633c565dca6b59b00513c797f78b144173#diff-430f630553c33723423ecadaccd9f19bL81

File: src/components/select/filterable-select/index.d.ts [hunk 3]
```diff
@@ -48,6 +49,10 @@ export interface FilterableSelectProps {
   onBlur?: () => void;
   /** Callback function for when the key is pressed when focused on Select Textbox. */
   onKeyDown?: () => void;
+  /** True for default text button or a Button Component to be rendered */
+  listActionButton?: boolean | typeof Button;
```
Comment (sjohnson-sage) [inline src/components/select/filterable-select/index.d.ts:53]: This doesn't have the option to be boolean in the propTypes in the component file

File: src/components/select/list-action-button/list-action-button.style.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import styled from 'styled-components';
+import { baseTheme } from '../../../style/themes';
+import StyledButton from '../../button/button.style';
+
+const StyledListActionButtonWrapper = styled.div`
+  padding-top: 8px;
```
Comment (sjohnson-sage) [inline src/components/select/list-action-button/list-action-button.style.js:6]: Should we be using the theme multiplier here rather than hardcoding `8px`?

**Review by sjohnson-sage — APPROVED** [review]

**Review by sjohnson-sage — APPROVED** [review]

**Review by dawidzarzycki — APPROVED** [review]

**Review by DlgSHi — DISMISSED** [review]

File: src/components/select/list-action-button/list-action-button.component.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    }
+  );
+
+  function renderListActionButton() {
+    if (!listActionButton || listActionButton === true) {
+      return (
+        <Button
+          forwardRef={ ref }
+          onClick={ onListAction }
+          iconType='add'
```
Comment (sjohnson-sage) [inline src/components/select/list-action-button/list-action-button.component.js:21]: Will this always be the `add` icon?

Comment (grabkowski) [reply src/components/select/list-action-button/list-action-button.component.js:21]: yes, as a default. You could provide your own button with your own icon.

Comment (sjohnson-sage) [reply src/components/select/list-action-button/list-action-button.component.js:21]: Apologies I didn't read that properly this time, it's clearly a default!

**Review by sjohnson-sage — DISMISSED** [review]

**Review by dawidzarzycki — APPROVED** [review]

**Review by sjohnson-sage — APPROVED** [review]

**Review by dawidzarzycki — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Pull request merged**